### DISCUSSION TITLE
Possible to access the x-data outside of the scope in alpinejs

### DISCUSSION THREAD

**Discussion by Waleedviews** [alpinejs/alpine#3384] [1. Help] [answered]
i have a order list of the data, and each `list item` has popup modal attached, when user click on any list item it should show `x-data` value in the popup

```
<li x-data="{ friends: {{ $value['friends'] }}, mediId: {{$value['medid']}}  }"
                            x-on:click.prevent="$dispatch('open-modal', 'confirm-result')"</li>
<li x-data="{ friends: {{ $value['friends'] }}, mediId: {{$value['medid']}}  }"
                            x-on:click.prevent="$dispatch('open-modal', 'confirm-result')"</li>
````

below is my popup modal component i try this but it doesn't work `<h1 x-text="friends"></h1>`

```
<x-modal  maxWidth="lg" name="confirm-result" focusable>
        <form method="post" action="" class="p-6">
<h1 x-text="friends"></h1>
</form>
    </x-modal>
```

any help with this?

**Comment by ekwoka** [comment]
the modal is inside the `li`? if so, then you don't need the dispatch at all...

and you shouldn't have click listeners on non-focusable elements like buttons

**Reply by Waleedviews** [reply]
the model is out side the `li`, actually `li` is in for each loop, so i want when someone click on `li` element it should open the popup and shows the `li` data inside the modal.

**Reply by ekwoka** [reply]
so the modal is within some shared parent component with the LI?

I'd just make a function for it on the parent that you can pass the data to

**Comment by Waleedviews** [comment, marked as answer]
that's how it's listed.

```
<section>
<ul>
@foreach($friendlist-> $friend)
<li x-data="{ friends: {{ $friend['friends'] }}, mediId: {{$friend['medid']}}  }"
                            x-on:click.prevent="$dispatch('open-modal', 'confirm-result')"</li>
@endforeach
</ul>
</section>

<x-modal  maxWidth="lg" name="confirm-result" focusable>
        <form method="post" action="" class="p-6">
<h1 x-text="friends"></h1>
</form>
    </x-modal>
```

i am new to Alpine and basically learning it. So it would be great if you could help, somewhere i read it could be resolve with the help of the dispatch.

**Reply by ekwoka** [reply]
Dispatch is MOSTLY useless, tbh.

In this case, you can make a parent `x-data` to wrap and handle it.


```html
<div x-data="{ isOpen: false, data: null }" x-effect="isOpen = Boolean(data)">
   <ul>
    <li x-data="{ friends: {{$friend['friends']}} }">
      <button @click="data = friends"> Click me </button>
    </li>
  </ul>
  <x-modal isOpen="isOpen" x-text="JSON.stringify(data)"></x-modal>
</div>
```
I don't actually know what x-modal is to give you more clarity on how it works.

**Reply by Waleedviews** [reply]
let me test, quick question, 
**what does ```Dispatch``` actually do then?**

**Reply by ekwoka** [reply]
It dispatches an event from the current element.

Most often, I find that reaching for dispatch or refs often points to bad application design.

they can mainly have use cases when needing to interface with third party code that expects to be passed elements or listens for events.

**Reply by Waleedviews** [reply]
thank you so much, i try your solution it's working, however when i try it on the modal dialog it doesn't working, can you look at the code please?

```
<div class="md:grid md:grid-cols-1 md:gap-x-14 gap-8 max-w-2xl mx-auto pt-14" x-data="{ modelOpen: false }">
<ul>
    <li x-data="{ friends: {{$friend['friends']}} }">
      <button  @click="modelOpen =!modelOpen; data = friends"> Click me </button>
    </li>
    <li x-data="{ friends: {{$friend['friends']}} }">
      <button @click="modelOpen =!modelOpen; data = friends"> Click me </button>
    </li>
    <li x-data="{ friends: {{$friend['friends']}} }">
      <button @click="modelOpen =!modelOpen; data = friends"> Click me </button>
    </li>
    </ul>


     <div x-show="modelOpen" class="fixed inset-0 z-50 overflow-y-auto" aria-labelledby="modal-title" role="dialog" aria-modal="true">
                    <div class="flex items-end justify-center min-h-screen px-4 text-center md:items-center sm:block sm:p-0">
                        <div x-cloak @click="modelOpen = false" x-show="modelOpen"
                             x-transition:enter="transition ease-out duration-300 transform"
                             x-transition:enter-start="opacity-0"
                             x-transition:enter-end="opacity-100"
                             x-transition:leave="transition ease-in duration-200 transform"
                             x-transition:leave-start="opacity-100"
                             x-transition:leave-end="opacity-0"
                             class="fixed inset-0 transition-opacity bg-gray-500 bg-opacity-40" aria-hidden="true"
                        ></div>

                        <div x-cloak x-show="modelOpen"
                             x-transition:enter="transition ease-out duration-300 transform"
                             x-transition:enter-start="opacity-0 translate-y-4 sm:translate-y-0 sm:scale-95"
                             x-transition:enter-end="opacity-100 translate-y-0 sm:scale-100"
                             x-transition:leave="transition ease-in duration-200 transform"
                             x-transition:leave-start="opacity-100 translate-y-0 sm:scale-100"
                             x-transition:leave-end="opacity-0 translate-y-4 sm:translate-y-0 sm:scale-95"
                             class="inline-block w-full max-w-xl p-8 my-20 overflow-hidden text-left transition-all transform bg-white rounded-lg shadow-xl 2xl:max-w-2xl"
                        >
                            <div class="flex items-center justify-between space-x-4">
                                <h1 class="text-xl font-medium text-gray-800 " x-text="JSON.stringify(data)"></h1>
                            </div>
                        </div>
                    </div>
                </div>
 </div>
```

**Reply by ekwoka** [reply]
What does `not work` mean? you also need to set the key `data` on the parent component, otherwise it will be added to the closest component.

**Reply by ekwoka** [reply]
I also recommend using an effect or a getter/setter on the `isOpen` part so you only need to set data and not handle both things.

**Reply by ekwoka** [reply]
I'd also recommend putting the modal inside the overlay, and jsut having the modal itself have an `@click.stop` to prevent clicks in the modal from spilling to the overlay.

**Reply by Waleedviews** [reply]
actually it worked quite good when you don't use modal dialog, when i access the data in modal dialog it give me an error in console.
```Alpine Expression Error: Can't find variable: data```

**Reply by ekwoka** [reply]
Yeah, since you didn't define `data` on the parent data component.

**Reply by Waleedviews** [reply]
Sorry you are correct. let me re try.  thank you so much for all the help.

**Reply by Waleedviews** [reply]
it's working,  thank you so much, last thing 

```
x-data="{
friends: 'zura,
mediId: '1234'
}"
```

if i have multiple data in x-data, how can i pass it in one go, instead of writing it like this 

```
@click="modelOpen =!modelOpen; data = friends;  data = mediId"
```

**Reply by ekwoka** [reply]
Just normal javascript assignment

```js
data = { friends, mediId }
```

**Reply by Waleedviews** [reply]
thank you, all good. thanks for your time.
